### PR TITLE
Add certificate verification and hostname validation for clients

### PR DESCRIPTION
**Pull request by arnoldgeels**
When I used the excellent oat++ library to make a call, I noticed that the certificate from the server was not checked against a list of trusted roots, and that the server name was not matched. Certificate verification and hostname validation protect against man-in-the-middle attacks. I modified the code to add these features, then decided to submit it.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,10 +32,14 @@ auto connectionProvider = oatpp::openssl::server::ConnectionProvider::createShar
...
 #include "oatpp-openssl/Config.hpp"
+#include "oatpp-openssl/configurer/TrustStore.hpp"
 
 ...
 
-auto config = oatpp::openssl::Config::createShared();
+const char* trust = "path/to/truststore";
+
+auto config = oatpp::openssl::Config::createDefaultClientConfigShared();
+config->addContextConfigurer(std::make_shared<oatpp::openssl::configurer::TrustStore>(trust, nullptr));
```
Comment (arnoldgeels) [inline README.md:42]: Show how to add a trust store in the readme

File: src/oatpp-openssl/Connection.cpp [hunk 2]
```diff
@@ -71,10 +68,8 @@ void Connection::ConnectionContext::init() {
         case SSL_ERROR_WANT_WRITE:
           break;
         default:
-          //std::string errStr = std::string(ERR_error_string(ERR_get_error(), nullptr));
-          //OATPP_LOGE("[oatpp::openssl::Connection::ConnectionContext::init()]", "Error. Handshake failed. code=%d, msg='%s'", err, errStr.c_str());
-          //throw std::runtime_error("[oatpp::openssl::Connection::ConnectionContext::init()]: Error. Handshake failed. " + errStr);
-          return;
+          ErrorStack::logErrors("[oatpp::openssl::Connection::ConnectionContext::init()]");
+          throw std::runtime_error("[oatpp::openssl::Connection::ConnectionContext::init()]: Error. Handshake failed.");
```
Comment (arnoldgeels) [inline src/oatpp-openssl/Connection.cpp:72]: If the certificate isn't valid or the host name does not match, this is where the failure will occur.  Log the OpenSSL error stack (there will typically be more than one error message)

File: src/oatpp-openssl/Connection.cpp [hunk 3]
```diff
@@ -119,9 +114,8 @@ async::CoroutineStarter Connection::ConnectionContext::initAsync() {
         case SSL_ERROR_WANT_WRITE:
           break;
         default:
-          //OATPP_LOGE("[oatpp::openssl::Connection::ConnectionContext::initAsync()]", "Error. Handshake failed. err=%d, errno=%d", err, er);
-          //throw std::runtime_error("[oatpp::openssl::Connection::ConnectionContext::initAsync()]: Error. Handshake failed.");
-          return finish();
+          ErrorStack::logErrors("[oatpp::openssl::Connection::ConnectionContext::initAsync()]");
+          return error<Error>("[oatpp::openssl::Connection::ConnectionContext::initAsync()]: Error. Handshake failed.");
```
Comment (arnoldgeels) [inline src/oatpp-openssl/Connection.cpp:118]: Same for async.

File: src/oatpp-openssl/client/ConnectionProvider.cpp [hunk 4]
```diff
@@ -125,6 +125,7 @@ provider::ResourceHandle<data::stream::IOStream> ConnectionProvider::get(){
   SSL_set_mode(ssl, SSL_MODE_ENABLE_PARTIAL_WRITE);
   SSL_set_connect_state(ssl);
   SSL_set_tlsext_host_name(ssl, host->c_str());
+  SSL_set1_host(ssl, host->c_str());
```
Comment (arnoldgeels) [inline src/oatpp-openssl/client/ConnectionProvider.cpp:128]: Set the host name for validation.  This triggers the validation of the name against the SAN in the certificate.

File: test/oatpp-openssl/FullAsyncClientTest.cpp [hunk 5]
```diff
@@ -117,6 +118,8 @@ class TestComponent {
     }
 
     auto config = oatpp::openssl::Config::createDefaultClientConfigShared();
+    OATPP_LOGD("oatpp::openssl::Config", "trust='%s'", CERT_CRT_PATH);
+    config->addContextConfigurer(std::make_shared<oatpp::openssl::configurer::TrustStore>(CERT_CRT_PATH, nullptr));
```
Comment (arnoldgeels) [inline test/oatpp-openssl/FullAsyncClientTest.cpp:122]: Set up the tests to check the certificate and host name.

File: src/oatpp-openssl/configurer/TrustStore.cpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  : m_cafile(file), m_cadir(dir)
+{}
+
+void TrustStore::configure(SSL_CTX *ctx) {
+  if (SSL_CTX_load_verify_locations(ctx, m_cafile?m_cafile->c_str():nullptr, m_cadir?m_cadir->c_str():nullptr) <= 0) {
+    ErrorStack::logErrors("[oatpp::openssl::configurer::TrustStore::configure()]");
+    throw std::runtime_error("Call to 'SSL_CTX_load_verify_locations' failed.");
+  }
+  SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, nullptr);
+}
```
Comment (arnoldgeels) [inline src/oatpp-openssl/configurer/TrustStore.cpp:42]: `SSL_CTX_load_verify_locations()` will give OpenSSL the list of trusted certificates to use for verification.
Actually turn on verification using `SSL_VERIFY_PEER`.

File: test/oatpp-openssl/FullAsyncClientTest.cpp [hunk 7]
```diff
@@ -162,6 +165,7 @@ class ClientCoroutine_getRootAsync : public oatpp::async::Coroutine<ClientCorout
   }
 
   Action handleError(Error* error) override {
+    SUCCESS_COUNTER = -2;
```
Comment (arnoldgeels) [inline test/oatpp-openssl/FullAsyncClientTest.cpp:168]: This is not strictly part of the verification/validation, but the test would time out if the handshake failed.  Propagate the failure so the test fails quickly.

File: utility/cert/oatpp.cnf [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+C = UA
+ST = Kiev
+L = City
+O = oatpp
+OU = hq
+CN = localhost
+emailAddress = lganzzzo@gmail.com
+
+[ext]
+subjectAltName = DNS:localhost,DNS:virtualhost
```
Comment (arnoldgeels) [inline utility/cert/oatpp.cnf:22]: This is the trick to make the unit tests work.  Normally, the SAN is the fully qualified hostname.

File: utility/cert/test_cert.crt [hunk 9]
```diff
@@ -1,32 +1,33 @@
...
+QmAzWQnYwO/dICma31wG66ihBbuXD+7kktCHbl690HweQgPorJjnjj2VSc0vpYL0
+o4sspXVOZqrUjAh4PKj4/Ns2epIACOzlXKMdYmnEzgll4NJFylZjUdb5J4psKP8p
+E9DQ20KNRd7wZmtC+9Vrchi7Jsqb+HWvLGhaCy0lY4IabkD8buyRjoiFnRbhEp/N
+bnx+GWSFtbyVMs48w9OJsTyhaxlmZxTA12HltG+v+RQhOXnYASHJ0EKmisO0/uBL
+VR/IwBPW0TWNDAExzOqE35yh3amwmq/Axzqucgb7YZPlJ580knoz2Tr1EwBgwFXq
+UWN3ivOli0s6SLl2vMZGRiZk6VXiEYVUK7zNgk7Pw4ACWhk5wr/9KRmWaQAU6IDk
+bNg6YCEKSjiGi8I1kYXfxitP71VPGYTw2WBBIOOKSkb7a/KY/q6bg0K0MkdZACRv
+CgtTF6zJiNr8Dp59hnQ/ljNDbvGrovjiEqK+4+rCs1vyujwROOSiaLP4/5Qo23so
+XZd12HY2Arb9AcxrGmpeu1+H67sWkWsv82JRVQGpHp1hDL9gCvF22ToUttzKwZ7q
+AX6znd7crITVR82N7yGEZhqk9sHu
```
Comment (arnoldgeels) [inline utility/cert/test_cert.crt:32]: For the unit test, this file now serves as both the certificate for the server and as the trust store for the client.  Normally, these would be separate files.

**Review by lganzzzo — APPROVED** [review]

**Comment by lganzzzo** [comment]
This is a great PR,
Thank you @arnoldgeels!
Merged

**Pull request merged**